### PR TITLE
Et2f main

### PR DESCRIPTION
**Pull request by robertavram**
making a pr to observe progress

### CODE REVIEW

File: EquiTrack/et2f/views.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        dataset = TravelListExporter().export(queryset)
+
+        response = HttpResponse(content_type='text/csv')
+        response['Content-Disposition'] = 'attachment; filename="ModelExportPartners.csv"'
+        response.write(dataset.csv)
+
+        return response
+
+    # State changes
+    submit_for_approval = state_transition('submit_for_approval')
```
Comment (robertavram) [inline EquiTrack/et2f/views.py]: technically anti-pattern.. I think it's beautiful.. "tastefully offensive" since I can't propose a better option at the moment... let's keep it... but keep in mind that we might have to change in the future.

File: EquiTrack/et2f/views.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        return self.get_paginated_response(serializer.data)
+
+    def retrieve(self, request, *args, **kwargs):
+        instance = self.get_object()
+        serializer = TravelDetailsSerializer(instance)
+        return Response(serializer.data, status.HTTP_200_OK)
+    
+    def update(self, request, *args, **kwargs):
+        partial = kwargs.pop('partial', False)
+        transition_name = kwargs.pop('transition_name', None)
```
Comment (robertavram) [inline EquiTrack/et2f/views.py]: theoretically you could set the status in the data by transition name... then you really don;t have to pass any data on the transition calls.

File: EquiTrack/et2f/views.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    mark_as_done = state_transition('mark_as_done')
+    submit_certificate = state_transition('submit_certificate')
+    approve_cetificate = state_transition('approve_cetificate')
+    reject_certificate = state_transition('reject_certificate')
+    mark_as_certified = state_transition('submit_for_approval')
+    mark_as_completed = state_transition('mark_as_completed')
+
+
+class StaticDataViewSet(mixins.ListModelMixin,
+                        viewsets.GenericViewSet):
```
Comment (robertavram) [inline EquiTrack/et2f/views.py]: no need for viewset here.. let's go to GenericApiView

File: EquiTrack/EquiTrack/factories.py [hunk 4]
```diff
@@ -1,20 +1,23 @@
 """
 Model factories used for generating models dynamically for tests
 """
+
 __author__ = 'jcranwellward'
```
Comment (robertavram) [inline EquiTrack/EquiTrack/factories.py]: you can safely remove line 5

File: EquiTrack/EquiTrack/settings/base.py [hunk 5]
```diff
@@ -333,13 +333,15 @@
 # See: https://docs.djangoproject.com/en/dev/ref/settings/#middleware-classes
 MIDDLEWARE_CLASSES = (
     # Default Django middleware.
+    'et2f.middlewares.FixRestFrameworkFuckupMiddleware',
```
Comment (tejfel) [inline EquiTrack/EquiTrack/settings/base.py]: Fixed.

File: EquiTrack/et2f/__init__.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
+from __future__ import unicode_literals
+
+from django.utils.translation import ugettext_lazy as _
+
+
```
Comment (tejfel) [inline EquiTrack/et2f/__init__.py]: The reason why I put here because if it's used in a migration for example, then the imports are not importing anything what should not be imported. But I'll move if you want.

File: EquiTrack/et2f/middlewares.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
+
+from django.contrib.auth import authenticate
+
+from et2f import PULI_USER_USERNAME, PULI_USER_PASSWORD
+
+
+class FixRestFrameworkFuckupMiddleware(object):
```
Comment (tejfel) [inline EquiTrack/et2f/middlewares.py]: Fixed

File: EquiTrack/et2f/models.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    arrival_date = models.DateTimeField()
+    dsa_region = models.ForeignKey('DSARegion', related_name='+')
+    overnight_travel = models.BooleanField(default=False)
+    mode_of_travel = models.CharField(max_length=255)
+    airlines = models.ManyToManyField('AirlineCompany', related_name='+')
+
+
+class Expense(models.Model):
+    travel = models.ForeignKey('Travel', related_name='expenses')
+    type = models.CharField(max_length=64)
```
Comment (tejfel) [inline EquiTrack/et2f/models.py]: Fixed.

File: EquiTrack/et2f/views.py (hunk 1 above)
Comment (tejfel) [inline EquiTrack/et2f/views.py]: This was the cleanest what I was able to figure out. If you find any better let me know and I'll change it.

File: EquiTrack/et2f/views.py (hunk 3 above)
Comment (robertavram) [inline EquiTrack/et2f/views.py]: yes .. please do:)

**Comment by robertavram** [comment]
@tejfel can you update from develop please.. many changes will be coming soon

**Comment by robertavram** [comment]
@n1207n could you look through this pr and review as much as you can?

**Comment by n1207n** [comment]
@robertavram Sure thing. I'll take a look ;)

File: EquiTrack/et2f/serializers/__init__.py [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+                model = self._fields[attr_name].Meta.model
+                self.create_related_models(model, [related_data], travel=self.instance)
+            elif related_data is None and related:
+                related.delete()
+
+
+class TravelListSerializer(TravelDetailsSerializer):
+    traveler = serializers.CharField(source='traveler.get_full_name')
+
+    class Meta(TravelDetailsSerializer.Meta):
```
Comment (n1207n) [inline EquiTrack/et2f/serializers/__init__.py]: Django REST Framework officially recommends NOT to inherit the inner Meta class from another Serializer. Define it separately and explicitly.

File: EquiTrack/et2f/serializers/__init__.py [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+
+class CloneParameterSerializer(serializers.Serializer):
+    traveler = serializers.PrimaryKeyRelatedField(queryset=get_user_model().objects.all())
+
+    class Meta:
+        fields = ('traveler',)
+
+
+class CloneOutputSerializer(TravelDetailsSerializer):
+    class Meta(TravelDetailsSerializer.Meta):
```
Comment (n1207n) [inline EquiTrack/et2f/serializers/__init__.py]: Same goes for meta class inheritance from above comment.

File: EquiTrack/et2f/tests/test_views.py [hunk 11]
```diff
@@ -0,0 +1,379 @@
...
+
+        second_traveler = UserFactory()
+        data = {'traveler': second_traveler.id}
+        response = self.forced_auth_req('post', '/api/et2f/travels/{}/duplicate_travel/'.format(new_travel_id),
+                                        data=data, user=self.unicef_staff)
+        response_json = json.loads(response.rendered_content)
+        self.assertNotEqual(response_json['id'], new_travel_id)
+        self.assertEqual(response_json, {})
+
+# t = Travel.objects.get(id=response_json['id'])
```
Comment (n1207n) [inline EquiTrack/et2f/tests/test_views.py]: If we're not using these test code, then we should remove them.

File: EquiTrack/et2f/tests/test_views.py [hunk 12]
```diff
@@ -0,0 +1,379 @@
...
+                                  value=True)
+                    permissions.append(TravelPermission(**kwargs))
+
+    TravelPermission.objects.bulk_create(permissions)
+
+    response = self.forced_auth_req('get', '/api/et2f/permission_matrix/', user=self.unicef_staff)
+    response_json = json.loads(response.rendered_content)
+    from pprint import pformat
+
+    with open('ki.json', 'w') as out:
```
Comment (n1207n) [inline EquiTrack/et2f/tests/test_views.py]: This test case does not have any assertion or checks at all. Moreover, it's gonna create ki.json at current directory which `manage.py test` command runs.

Is there any reason behind this action?

File: EquiTrack/et2f/tests/test_views.py [hunk 13]
```diff
@@ -0,0 +1,379 @@
...
+            'file': fakefile}
+    response = self.forced_auth_req('post', '/api/et2f/travels/{}/attachments/'.format(travel.id), data=data,
+                                    user=self.unicef_staff, request_format='multipart')
+    response_json = json.loads(response.rendered_content)
+    # self.assertEqual(response_json, {})
+
+    response = self.forced_auth_req('delete',
+                                    '/api/et2f/travels/{}/attachments/{}/'.format(travel.id, response_json['id']),
+                                    user=self.unicef_staff)
+    self.assertEqual(response.status_code, 0)
```
Comment (n1207n) [inline EquiTrack/et2f/tests/test_views.py]: I have not seen HTTP status code 0 before ;)

Is this testing 200 OK or 204 No-Content?

File: EquiTrack/et2f/tests/test_views.py [hunk 14]
```diff
@@ -0,0 +1,379 @@
...
+    attachment.file.save('fake.txt', fakefile)
+    fakefile.seek(0)
+
+    data = {'name': 'second',
+            'type': 'something',
+            'file': fakefile}
+    response = self.forced_auth_req('post', '/api/et2f/travels/{}/attachments/'.format(travel.id), data=data,
+                                    user=self.unicef_staff, request_format='multipart')
+    response_json = json.loads(response.rendered_content)
+    # self.assertEqual(response_json, {})
```
Comment (n1207n) [inline EquiTrack/et2f/tests/test_views.py]: We should still check response code here even if DELETE method will be called after.

File: EquiTrack/et2f/tests/test_views.py [hunk 15]
```diff
@@ -0,0 +1,379 @@
...
+                                    '/api/et2f/travels/{}/attachments/{}/'.format(travel.id, response_json['id']),
+                                    user=self.unicef_staff)
+    self.assertEqual(response.status_code, 0)
+
+
+def test_duplication(self):
+    response = self.forced_auth_req('post', '/api/et2f/travels/{}/add_driver/'.format(self.travel.id),
+                                    user=self.unicef_staff)
+    response_json = json.loads(response.rendered_content)
+    self.assertEqual(response_json, {})
```
Comment (n1207n) [inline EquiTrack/et2f/tests/test_views.py]: Check the response code here. Empty response doesn't tell you anything about the REST API operation.

File: EquiTrack/et2f/tests/test_views.py [hunk 16]
```diff
@@ -0,0 +1,379 @@
...
+                "purpose": None,
+                "status": "submitted",
+                "mode_of_travel": [],
+                "estimated_travel_cost": "0.0000",
+                "currency": currency.id}
+
+        response = self.forced_auth_req('post', '/api/et2f/travels/', data=data,
+                                        user=self.unicef_staff)
+        response_json = json.loads(response.rendered_content)
+        self.assertEqual(response_json, {})
```
Comment (n1207n) [inline EquiTrack/et2f/tests/test_views.py]: Check the HTTP response code here.

File: EquiTrack/et2f/tests/test_views.py [hunk 17]
```diff
@@ -0,0 +1,379 @@
...
+        new_travel_id = response_json['id']
+
+        second_traveler = UserFactory()
+        data = {'traveler': second_traveler.id}
+        response = self.forced_auth_req('post', '/api/et2f/travels/{}/duplicate_travel/'.format(new_travel_id),
+                                        data=data, user=self.unicef_staff)
+        response_json = json.loads(response.rendered_content)
+        self.assertNotEqual(response_json['id'], new_travel_id)
+        self.assertEqual(response_json, {})
```
Comment (n1207n) [inline EquiTrack/et2f/tests/test_views.py]: Check the HTTP response code here.

File: EquiTrack/et2f/tests/test_views.py [hunk 18]
```diff
@@ -0,0 +1,379 @@
...
+# response = self.forced_auth_req('patch', '/api/et2f/travels/{}/'.format(response_json['id']),
+#                                 data=response_json,
+#                                 user=self.unicef_staff)
+# response_json = json.loads(response.rendered_content)
+# self.assertEqual(response_json, {})
+
+def test_static_data_endpoint(self):
+    response = self.forced_auth_req('get', '/api/et2f/static_data/', user=self.unicef_staff)
+    response_json = json.loads(response.rendered_content)
+    self.assertEqual(response_json, {})
```
Comment (n1207n) [inline EquiTrack/et2f/tests/test_views.py]: Is empty response considered a successful response here for static data GET method? If so, we should check response code for OK.

File: EquiTrack/et2f/tests/test_views.py [hunk 19]
```diff
@@ -0,0 +1,379 @@
...
+
+
+def test_curent_user_endpoint(self):
+    response = self.forced_auth_req('get', '/api/et2f/me/', user=self.unicef_staff)
+    response_json = json.loads(response.rendered_content)
+    self.assertEqual(response_json, {})
```
Comment (n1207n) [inline EquiTrack/et2f/tests/test_views.py]: Same goes for empty response and response code comment above.

File: EquiTrack/et2f/tests/test_views.py [hunk 20]
```diff
@@ -0,0 +1,379 @@
...
+
+
+def test_payload(self):
+    TravelPermission.objects.create(name='afds', code='can_see_travel_status', user_type='God', status='planned')
+
+    travel = TravelFactory()
+    response = self.forced_auth_req('get', '/api/et2f/travels/{}/'.format(travel.id), user=self.unicef_staff)
+    self.assertEqual(json.loads(response.rendered_content), {})
```
Comment (n1207n) [inline EquiTrack/et2f/tests/test_views.py]: Same goes for empty response and response code comment above.

**Comment by n1207n** [comment]
I took a review on this PR and it looks pretty solid in general. However, there are some improvements to be made according to my experience. Take a look and let me know if you have any comments!

File: EquiTrack/EquiTrack/mixins.py [hunk 21]
```diff
@@ -130,7 +130,11 @@ def process_request(self, request):
 class EtoolsTokenAuthentication(TokenAuthentication):
 
     def authenticate(self, request):
-        user, token = super(EtoolsTokenAuthentication, self).authenticate(request)
+        super_return = super(EtoolsTokenAuthentication, self).authenticate(request)
+        if super_return is None:
```
Comment (tejfel) [inline EquiTrack/EquiTrack/mixins.py]: In my opinion it's more readable like this and the speed improvement would not be significant. But I can change if you insist.

Comment (n1207n) [inline EquiTrack/EquiTrack/mixins.py]: In fact, evaluating None value without `is` keyword is considered more Pythonic. It's simpler in syntax and not redundant.

Again, `is` keyword is reserved to compare the type identity. If we're simply trying to see if an object has None value, then let the variable evaluates itself.

File: EquiTrack/et2f/views.py [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+from users.models import Office, Section
+from .exports import TravelListExporter
+from .models import Travel
+from .serializers import TravelListSerializer, TravelDetailsSerializer, TravelListParameterSerializer
+
+
+class PageNumberPagination(_PageNumberPagination):
+    page_size = 10
+    page_size_query_param = 'page_size'
+
```
Comment (tejfel) [inline EquiTrack/et2f/views.py]: Check

File: EquiTrack/et2f/views.py [hunk 23]
```diff
@@ -0,0 +1,163 @@
...
+from locations.models import Location
+from partners.models import PartnerOrganization, PCA
+from reports.models import Result
+from users.models import Office, Section
+from .exports import TravelListExporter
+from .models import Travel
+from .serializers import TravelListSerializer, TravelDetailsSerializer, TravelListParameterSerializer
+
+
+class PageNumberPagination(_PageNumberPagination):
```
Comment (tejfel) [inline EquiTrack/et2f/views.py]: Check

File: EquiTrack/et2f/views.py [hunk 24]
```diff
@@ -0,0 +1,163 @@
...
+
+from collections import OrderedDict
+
+from django.contrib.auth import get_user_model
+from django.db.models.query_utils import Q
+from django.http.response import HttpResponse
+from rest_framework import viewsets, mixins, status
+from rest_framework.decorators import list_route, detail_route
+from rest_framework.permissions import IsAdminUser
+from rest_framework.pagination import PageNumberPagination as _PageNumberPagination
```
Comment (tejfel) [inline EquiTrack/et2f/views.py]: Check

File: EquiTrack/EquiTrack/factories.py (hunk 4 above)
Comment (tejfel) [inline EquiTrack/EquiTrack/factories.py]: Check

File: EquiTrack/et2f/serializers/__init__.py (hunk 9 above)
Comment (tejfel) [inline EquiTrack/et2f/serializers/__init__.py]: Check

File: EquiTrack/et2f/models.py [hunk 25]
```diff
@@ -0,0 +1,289 @@
...
+        return self.approved_at
+
+    @property
+    def cost_summary(self):
+        calculator = CostSummaryCalculator(self)
+        return calculator.calculate_cost_summary()
+
+    # State machine transitions
+    def check_completion_conditions(self):
+        if self.status is TripStatus.SUBMITTED and not self.international_travel:
```
Comment (tejfel) [inline EquiTrack/et2f/models.py]: Check

File: EquiTrack/et2f/helpers.py [hunk 26]
```diff
@@ -0,0 +1,203 @@
+from __future__ import unicode_literals
+
+from datetime import timedelta, datetime
+from decimal import Decimal as D
```
Comment (tejfel) [inline EquiTrack/et2f/helpers.py]: Check

File: EquiTrack/et2f/helpers.py [hunk 27]
```diff
@@ -0,0 +1,203 @@
...
+
+class PermissionMatrix(object):
+    def __init__(self, travel, user):
+        self.travel = travel
+        self.user = user
+        self._user_type = self.get_user_type()
+        self._permission_dict = self.get_permission_dict()
+
+    def get_user_type(self):
+        return UserTypes.GOD
```
Comment (tejfel) [inline EquiTrack/et2f/helpers.py]: This is just for internal testing :)

File: EquiTrack/et2f/management/commands/_private.py [hunk 28]
```diff
@@ -0,0 +1,227 @@
...
+        self.set_up_representative()
+
+    def revoke_edit(self, qs):
+        num_revoked = qs.filter(permission_type=TravelPermission.EDIT).update(value=False)
+        self.log('{} permissions revoked'.format(num_revoked))
+
+    def grant_edit(self, qs):
+        num_revoked = qs.filter(permission_type=TravelPermission.EDIT).update(value=True)
+        self.log('{} permissions revoked'.format(num_revoked))
+    def revoke_view(self, qs):
```
Comment (tejfel) [inline EquiTrack/et2f/management/commands/_private.py]: Check

File: EquiTrack/et2f/management/commands/_private.py [hunk 29]
```diff
@@ -0,0 +1,227 @@
...
+                             TripStatus.REJECTED]
+
+        q = Q(status__in=status_where_hide) & self.get_related_q(['deductions', 'expenses'])
+        sub_qs = qs.filter(q)
+        self.revoke_view(sub_qs)
+        self.revoke_edit(sub_qs)
+
+        sub_qs = qs.filter(status__in=[TripStatus.SUBMITTED, TripStatus.CERTIFICATION_SUBMITTED])
+        self.revoke_edit(sub_qs)
+    def set_up_travel_administrator(self):
```
Comment (tejfel) [inline EquiTrack/et2f/management/commands/_private.py]: Check

File: EquiTrack/et2f/management/commands/_private.py [hunk 30]
```diff
@@ -0,0 +1,227 @@
...
+        qs = TravelPermission.objects.filter(user_type=UserTypes.TRAVEL_FOCAL_POINT)
+        self.revoke_edit(qs)
+
+        q = Q(status__in=status_where_edit) & self.get_related_q(fields_to_edit)
+        num_granted = qs.filter(q).update(value=True)
+        self.log('{} permissions granted'.format(num_granted))
+
+        sub_q = qs.filter(model='travel', field__in=['estimated_travel_cost', 'currency'])
+        self.grant_edit(sub_q)
+    def set_up_finance_focal_point(self):
```
Comment (tejfel) [inline EquiTrack/et2f/management/commands/_private.py]: Check

File: EquiTrack/et2f/management/commands/_private.py [hunk 31]
```diff
@@ -0,0 +1,227 @@
...
+        qs = TravelPermission.objects.filter(user_type=UserTypes.FINANCE_FOCAL_POINT)
+        self.revoke_edit(qs)
+
+        q = Q(status__in=status_where_edit) & self.get_related_q(fields_to_edit)
+        num_granted = qs.filter(q).update(value=True)
+        self.log('{} permissions granted'.format(num_granted))
+
+    def set_up_representative(self):
+        permissions = TravelPermission.objects.filter(user_type=UserTypes.REPRESENTATIVE)
+        permissions = permissions.filter(permission_type=TravelPermission.EDIT)
```
Comment (tejfel) [inline EquiTrack/et2f/management/commands/_private.py]: Check

File: EquiTrack/et2f/models.py [hunk 32]
```diff
@@ -0,0 +1,289 @@
...
+    ta_required = models.NullBooleanField(default=True, null=True, blank=True)
+    reference_number = models.CharField(max_length=12, default=make_reference_number)
+    hidden = models.BooleanField(default=False)
+    mode_of_travel = models.ManyToManyField('TravelType', related_name='+')
+    estimated_travel_cost = models.DecimalField(max_digits=20, decimal_places=4, default=0)
+    currency = models.ForeignKey('Currency', null=True, blank=True, related_name='+')
+
+    @property
+    def is_driver(self):
+        return False
```
Comment (tejfel) [inline EquiTrack/et2f/models.py]: Transformed into a field.

File: EquiTrack/et2f/filters.py [hunk 33]
```diff
@@ -0,0 +1,91 @@
...
+                q |= Q(**{constructed_field_name: search_str})
+            queryset = queryset.filter(q)
+
+        return queryset
+
+
+class ShowHiddenFilter(BaseFilterBackend):
+    def filter_queryset(self, request, queryset, view):
+        serializer = ShowHiddenFilterSerializer(data=request.GET)
+        serializer.is_valid()
```
Comment (tejfel) [inline EquiTrack/et2f/filters.py]: Check.

File: EquiTrack/et2f/filters.py [hunk 34]
```diff
@@ -0,0 +1,91 @@
...
+    FilterBoxFilterSerializer
+
+
+class SearchFilter(BaseFilterBackend):
+    _search_fields = ('id', 'reference_number', 'traveler__first_name', 'traveler__last_name', 'purpose',
+                      'section__name', 'office__name', 'supervisor__first_name', 'supervisor__last_name')
+
+    def filter_queryset(self, request, queryset, view):
+        serializer = SearchFilterSerializer(data=request.GET)
+        serializer.is_valid()
```
Comment (tejfel) [inline EquiTrack/et2f/filters.py]: Check

File: EquiTrack/et2f/filters.py [hunk 35]
```diff
@@ -0,0 +1,91 @@
...
+            q = Q(hidden=True) | Q(status=TripStatus.CANCELLED)
+            queryset = queryset.exclude(q)
+
+        return queryset
+
+
+class SortFilter(BaseFilterBackend):
+    def filter_queryset(self, request, queryset, view):
+        serializer = SortFilterSerializer(data=request.GET)
+        serializer.is_valid()
```
Comment (tejfel) [inline EquiTrack/et2f/filters.py]: Check

File: EquiTrack/et2f/filters.py [hunk 36]
```diff
@@ -0,0 +1,91 @@
...
+        return queryset.order_by(sort_by)
+
+
+class FilterBoxFilter(BaseFilterBackend):
+    """
+    Does the filtering based on the filter parameters coming from the frontend
+    """
+    def filter_queryset(self, request, queryset, view):
+        serializer = FilterBoxFilterSerializer(data=request.GET)
+        serializer.is_valid()
```
Comment (tejfel) [inline EquiTrack/et2f/filters.py]: Check

File: EquiTrack/EquiTrack/mixins.py (hunk 21 above)
Comment (tejfel) [inline EquiTrack/EquiTrack/mixins.py]: Check

File: EquiTrack/et2f/helpers.py [hunk 37]
```diff
@@ -166,14 +167,14 @@ def _do_the_cloning(self, new_traveler, fk_related, o2o_related):
         new_travel.reset_status()
         new_travel.save()
 
-        cloned_models = self._clone_related(fk_related, o2o_related)
+        cloned_models = self._clone_relateDecimal(fk_related, o2o_related)
```
Comment (n1207n) [inline EquiTrack/et2f/helpers.py]: I believe this got replaced mistakenly ;)

Comment (tejfel) [inline EquiTrack/et2f/helpers.py]: Yes :)

**Comment by n1207n** [comment]
Hey @tejfel, I have noticed your work on the permission matrix regarding its TravelPermission usage for according user type.

My impression on this so far is TravelPermission objects are created via management command as one-off effort, and permission_matrix REST API currently creates a dictionary structure to represent all permission entries for all user types.

I definitely agree with your one-off approach to create all travel permission related information, as they are fixed and very unlikely to be changed. However, I don't think your permission_matrix REST API is efficient. TravelPermission is considered a Django model class, therefore every REST API request from permission_matrix will involve DB queries and serialization effort.

What I would like to propose on top of your clever approach is to take what current permission_matrix REST API returns as the API response and save it as a Python constant dictionary. Basically, I'm proposing to take the final outcome from permission_matrix REST API as a fixture. No more DB query is involved. No more extra serialization is involved from DRF serializer, while using direct Response object from DRF to store the fixture data.

So tl;dr, Save the API response from permission_matrix REST API as a Python dictionary with constant naming convention. We can still leverage from your management command to create a temporary Python class instead in order to generate a new permission matrix fixture in the future.

Let me know how this suggestion sounds to you.

**Comment by robertavram** [comment]
@tejfel https://etools-dev.unicef.org/api/docs/ . there are some configuration issues on some of the apis...

**Comment by robertavram** [comment]
@tejfel add a todo on the permissions matrix generator... we should try to have that as static config file.. rather than generated in the model.. otherwise good. keep the generator, but the db shouldn't take an extra query for that as it will slow the apis, python native dict could store it as well.

**Comment by tejfel** [comment]
@n1207n @robertavram I added some caching to the endpoint yesterday already (0d0f85d). Should we use that or should I put the json somewhere in the folders and just serve it as is?

**Comment by n1207n** [comment]
@tejfel,

Django cache works fine too, as long as cache backend does not reboot or get flushed. I would stay away from storing it as json file as now Django needs to do deserialization every access.

The concern I have is when to invalidate permission matrix from the cache. Let's say hypothetically we update the permission matrix few years from now on. We have your management command to re generate or update TravelPermission models. Should we manually delete the cache when this happens?

**Comment by tejfel** [comment]
@n1207n Depending on the cache, yes, it should be cleaned. Or the management command can be extended to invalidate the cache on regeneration. Keep in mind that these data will be cached on the frontend too and I have no idea yet how to invalidate that.

**Comment by robertavram** [comment]
@tejfel please update from develop and make sure tests pass

File: EquiTrack/et2f/serializers/__init__.py [hunk 38]
```diff
@@ -0,0 +1,279 @@
...
+    def _readable_fields(self):
+        fields = super(PermissionBasedModelSerializer, self)._readable_fields
+        return [f for f in fields if self._can_read_field(f)]
+
+    def _check_permission(self, permission_type, field_name):
+        model_name = model_name = self.Meta.model.__name__.lower()
+
+        permission_matrix = self.context.get('permission_matrix')
+
+        if permission_matrix is None:
```
Comment (n1207n) [inline EquiTrack/et2f/serializers/__init__.py]: Use `if not permission_matrix` instead.

File: EquiTrack/et2f/models.py [hunk 39]
```diff
@@ -50,6 +50,27 @@ class ExpenseType(models.Model):
...
+        (TECHNICAL_SUPPORT, 'TECHNICAL SUPPORT'),
+        (MEETING, 'MEETING'),
+        (STAFF_DEVELOPMENT, 'STAFF DEVELOPMENT'),
+        (STAFF_ENTITLEMENT, 'STAFF ENTITLEMENT'),
+    )
+
+    name = models.CharField(max_length=32, choices=CHOICES)
+
+
+class ModeOfTravel(models.Model):
```
Comment (tejfel) [inline EquiTrack/et2f/models.py]: Rob agreed on that for now we will use the extra models for every choice data because the frontend can handle only the pk relations for now.

File: EquiTrack/et2f/migrations/0006_auto_20161201_1940.py [hunk 40]
```diff
@@ -0,0 +1,50 @@
...
+
+def reset_travel_type_relations(apps, schema_editor):
+    Travel = apps.get_model('et2f', 'Travel')
+    IteneraryItem = apps.get_model('et2f', 'IteneraryItem')
+    TravelActivity = apps.get_model('et2f', 'TravelActivity')
+    ModeOfTravel = apps.get_model('et2f', 'ModeOfTravel')
+    TravelType = apps.get_model('et2f', 'TravelType')
+
+    mode_of_travel = ModeOfTravel.objects.last()
+    travel_type = TravelType.objects.last()
```
Comment (tejfel) [inline EquiTrack/et2f/migrations/0006_auto_20161201_1940.py]: Removed.

File: EquiTrack/et2f/migrations/0006_auto_20161201_1940.py [hunk 41]
```diff
@@ -0,0 +1,50 @@
...
+    TravelType.objects.all().delete()
+    for choice in TravelTypeModel.CHOICES:
+        TravelType.objects.create(name=choice[0])
+
+
+def create_mode_of_travel_models(apps, schema_editor):
+    ModeOfTravel = apps.get_model('et2f', 'ModeOfTravel')
+
+    for choice in ModeOfTravelModel.CHOICES:
+        ModeOfTravel.objects.create(name=choice[0])
```
Comment (tejfel) [inline EquiTrack/et2f/migrations/0006_auto_20161201_1940.py]: ModeOfTravel was created in the migration before. Not possible to have any data in there.

File: EquiTrack/et2f/tests/test_travel_details.py [hunk 42]
```diff
@@ -32,3 +46,48 @@ def test_list_view(self):
...
+            def size(self):
+                return len(self)
+
+        fakefile = FakeFile('some stuff')
+        travel = TravelFactory()
+        attachment = TravelAttachment.objects.create(travel=travel,
+                                                     name='test_attachment',
+                                                     type='document')
+        attachment.file.save('fake.txt', fakefile)
+        fakefile.seek(0)
```
Comment (tejfel) [inline EquiTrack/et2f/tests/test_travel_details.py]: Check

File: EquiTrack/et2f/filters.py [hunk 43]
```diff
@@ -0,0 +1,95 @@
...
+    def filter_queryset(self, request, queryset, view):
+        serializer = FilterBoxFilterSerializer(data=request.GET)
+        if not serializer.is_valid():
+            return queryset
+        data = serializer.validated_data
+
+        # Construct a backend readable date
+        year = data.pop('year', None)
+        month = data.pop('month', None)
+        if year is not None:
```
Comment (tejfel) [inline EquiTrack/et2f/filters.py]: Check

File: EquiTrack/et2f/filters.py [hunk 44]
```diff
@@ -0,0 +1,95 @@
...
+            if month is not None:
```
Comment (tejfel) [inline EquiTrack/et2f/filters.py]: Check

File: EquiTrack/et2f/serializers/__init__.py (hunk 38 above)
Comment (tejfel) [inline EquiTrack/et2f/serializers/__init__.py]: Check

File: EquiTrack/EquiTrack/settings/base.py [hunk 45]
```diff
@@ -382,6 +383,7 @@
     'generic_links',
     'gunicorn',
     'post_office',
+    'django_fsm',
```
Comment (robertavram) [inline EquiTrack/EquiTrack/settings/base.py]: let's bring fsm in the tenant schema

File: EquiTrack/et2f/filters.py [hunk 46]
```diff
@@ -0,0 +1,95 @@
...
+
+        # TODO simon: figure out what to do with this
+        data.pop('cp_output', None)
+
+        return queryset.filter(**data)
+
+
+class TravelAttachmentFilter(BaseFilterBackend):
+    def filter_queryset(self, request, queryset, view):
+        return queryset.filter(travel__pk=view.kwargs['travel_pk'])
```
Comment (robertavram) [inline EquiTrack/et2f/filters.py]: comment that in

File: EquiTrack/et2f/management/commands/et2f_init.py [hunk 47]
```diff
@@ -0,0 +1,415 @@
...
+    def handle(self, *args, **options):
+        username = options['username']
+        password = options['password']
+        user = self._get_or_create_admin_user(username, password)
+        connection.set_tenant(user.profile.country)
+
+        self._load_currencies()
+        self._load_users()
+        self._load_airlines()
+        self._load_offices()
```
Comment (robertavram) [inline EquiTrack/et2f/management/commands/et2f_init.py]: for staging users, offices and partners should be disabled in this script.

File: dev_environ_setup_linux.md [hunk 48]
```diff
@@ -32,12 +32,21 @@ Step 3. Set Postgres user password to "postgres", and create PostGIS required ex
...
+```
+
 Step 4. Install Redis
 
 ```bash
 $ wget http://download.redis.io/redis-stable.tar.gz
 $ tar xvzf redis-stable.tar.gz
 $ cd redis-stable
+$ cd deps
+$ make hiredis lua jemalloc linenoise geohash-int
```
Comment (robertavram) [inline dev_environ_setup_linux.md]: comment in what this does

**Comment by robertavram** [comment]
@tejfel looks good just some minor minor minor changes requested...
@n1207n please do a last sanity check look through and let's merge this in.

File: EquiTrack/et2f/models.py [hunk 49]
```diff
@@ -194,6 +197,7 @@ def plan(self):
 
     @transition(status, source=[APPROVED], target=SENT_FOR_PAYMENT)
     def send_for_payment(self):
+        self.preserved_expenses = self.cost_summary['expenses_total']
```
Comment (tejfel) [inline EquiTrack/et2f/models.py]: It will always be there :)

File: dev_environ_setup_linux.md (hunk 48 above)
Comment (tejfel) [inline dev_environ_setup_linux.md]: Check

File: EquiTrack/et2f/management/commands/et2f_init.py (hunk 47 above)
Comment (tejfel) [inline EquiTrack/et2f/management/commands/et2f_init.py]: Flags added

File: EquiTrack/EquiTrack/settings/base.py (hunk 45 above)
Comment (tejfel) [inline EquiTrack/EquiTrack/settings/base.py]: fsm has no models. Does not really matter where it is.

Comment (robertavram) [inline EquiTrack/EquiTrack/settings/base.py]: no it doesn't from that point of view... just for organizational purposes. I'd rather have as many apps in the tenant apps by default unless they need to be in the public apps
